### DISCUSSION TITLE
Architecture compatibility?

### DISCUSSION THREAD

**Discussion by Slackadays** [leo-arch/clifm#192] [Q&A] [answered]
I noticed in the readme that Clifm works on x86 and ARM. Is this just the architectures it's been tested with, or does it really only support those two? If it only supports those two, then that's a problem.

**Comment by leo-arch** [comment, marked as answer]
Hi @Slackadays. No tests have been made beyond X86 and ARM, though it will most likely work on other architectures (as far as I know, there's nothing architecture-specific in **clifm**'s code).

**Comment by Slackadays** [comment]
You could add a bunch of build and test GH actions to verify that clifm works everywhere, kind of like Clipboard does. I noticed that the only real action right now verifies building on one platform. Does that sound like a reasonable proposition?

**Reply by leo-arch** [reply]
Absolutely! I've seen what you've done with Clipboard's builds, and it's pretty cool.

**Comment by leo-arch** [comment]
Hi @Slackadays! Forgot to say that we've been successfully building distro-specific packages for over a year (using the [OpenSUSE Build System](https://build.opensuse.org/package/show/home:archcrack/CliFM)) for these architectures: `x86_64`, `i585`, `armv7l`, `aarch64`, and `ppc64le`.

**Reply by Slackadays** [reply]
@leo-arch Nice! If those packages are building and are being used, then it might also be great to say that `clifm` works great on all of those architectures, or maybe just all generically.

**Reply by leo-arch** [reply]
All I can say is that packages for each of those architectures are building fine. But, since I have tested them myself on neither `aarch64` nor `ppc64le`, I cannot really guarantee they work as intended. Too bad.

**Reply by Slackadays** [reply]
You could set up a QEMU VM to test `ppc64le` and get an Oracle Cloud Free Tier VM for `aarch64`, so there's that.

**Reply by leo-arch** [reply]
Thanks for the suggestion! I'll give it a shot.

**Reply by Slackadays** [reply]
Hey @leo-arch! I'm happy to say that I just tested compiling clifm on one of my RISC-V SBCs, and it works great. However, there is some UI lag every time I do anything. I already checked the Performance section in the Wiki and that didn't help because the lag is only in the UI, not file access which happens to be almost instant here. You may want to optimize command handling, as that's probably where the bottleneck is. Also, my RISC-V board is a LicheePi with a single in-order core at 1GHz and 512MB of memory, so it's not unlike a typical PC from 20 years ago. This should be a fun thing to work on if you can get your hands on of these boards, as RISC-V is all open source and the LicheePi costs only $26.
![image](https://user-images.githubusercontent.com/69403549/228377572-95948d5f-9548-45bb-81b2-fa05f655d98e.png)

**Reply by leo-arch** [reply]
Hey Slack, that's really good news! Thanks for taking the time to test this thing.

As to the lag, what do you mean exactly? Is it a typing lag, or maybe screen refresh time, or something else?

If it is related to the low specs machine, I guess I could set up a VM and perform some tests there.

**Reply by Slackadays** [reply]
Here's a video I made that demonstrates the issue (notice how just listing everything out is almost instant)
[demo.webm](https://user-images.githubusercontent.com/69403549/228424870-a63643aa-306a-4ab1-85b2-94753b1abfb6.webm)

Also check out this cool link about UI lag in general: https://www.nngroup.com/articles/response-times-3-important-limits/

AND the exact place I got the LicheePi so you can have a go too! :) https://www.aliexpress.us/item/3256803554972410.html

**Reply by leo-arch** [reply]
There seems to be some lag (more specifically, when printing the prompt) after running shell commands isn't? I see no lag for internal commands.

**Reply by leo-arch** [reply]
I don't need to make any tests, I already know the cause: after running external command **clifm** updates the list of available binaries (needed to print command names suggestions), that is, it scans all files in **PATH**, which in slow systems might be noticeable.

**Reply by Slackadays** [reply]
Exactly, it's the UI rendering that lags after running a shell command, but not internal commands.

Edit: Would it be possible to cache the binaries available, or is this infeasible/incorrect for some systems?

**Reply by leo-arch** [reply]
I thanks for the links! I wish I could get one of those LicheePies too! Sounds great.

**Reply by leo-arch** [reply]
They are cached in memory indeed. The point is to have new binaries available immediately after they were created, for example, after installing a new package. Otherwise, you'd need to restart **clifm**. There should be a better approach (maybe an option to disable this feature for slow systems).

**Reply by Slackadays** [reply]
You could check if the last write times of all directories in PATH have changed since you last checked them, since any addition or deletion to bin directories will modify the parent's last write time. Then, all you would need to do is cache the times in addition to anything else.

**Reply by leo-arch** [reply]
Sounds great! I'll try to implement that. If it works, that would be a nice performance boost (maybe noticeable only in slow specs systems, but anyway)

**Reply by leo-arch** [reply]
Thanks Slack! Nice find, elegant solution.

**Reply by Slackadays** [reply]
Glad to help as always! :pray:

**Reply by leo-arch** [reply]
I got a first implementation working (78ceac032). The lag should be mostly gone now.

**Reply by leo-arch** [reply]
Hi @Slackadays. Have you tried this fix?

**Reply by Slackadays** [reply]
Hey @leo-arch! I just tried the latest commit on the LicheePi (using Ubuntu now due to built-in ethernet drivers) and it's a whole lot faster after doing each command. This new write time check seems to work really well :)
